File: src/java/EntityFX/src/EntityFX/Core/Generic/RandomMemoryBenchmarkBase.java
package EntityFX.Core.Generic;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.lang.reflect.Array;
import java.util.AbstractMap;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Random;
import java.util.AbstractMap.SimpleEntry;
import java.util.Map.Entry;
import java.util.concurrent.ThreadLocalRandom;

import javax.xml.crypto.dsig.keyinfo.KeyValue;

import EntityFX.Core.Writer;

public abstract class RandomMemoryBenchmarkBase<TResult> extends BenchmarkBase<TResult> {

    public RandomMemoryBenchmarkBase(Writer writer, boolean printToConsole) throws FileNotFoundException {
        super(writer, printToConsole);
        Iterrations = 500000;
        Ratio = 2;
    }

    public MemoryBenchmarkResult benchRandomMemory() throws IOException {
        SimpleEntry<Double, Integer>  int4k = measureArrayRandomRead(1024);
        output.writeLine("Random int 4k: %.2f MB/s", int4k.getKey());
        SimpleEntry<Double, Integer>  int512k = measureArrayRandomRead(131072);
        output.writeLine("Random int 512k: %.2f MB/s", int512k.getKey());
        SimpleEntry<Double, Integer>  int8m = measureArrayRandomRead(2097152);
        output.writeLine("Random int 8M: %.2f MB/s", int8m.getKey());

        SimpleEntry<Double, Long>  long4k = measureArrayRandomLongRead(1024);
        output.writeLine("Random long 4k: %.2f MB/s", long4k.getKey());
        SimpleEntry<Double, Long> long512k = measureArrayRandomLongRead(131072);
        output.writeLine("Random long 512k: %.2f MB/s", long512k.getKey());
        SimpleEntry<Double, Long> long8m = measureArrayRandomLongRead(2097152);
        output.writeLine("Random long 8M: %.2f MB/s", long8m.getKey());

        double avg = Arrays.stream(new double[] { int4k.getKey(), int512k.getKey(), int8m.getKey(), 
                long4k.getKey(), long512k.getKey(), long8m.getKey(), }).average().orElse(-1);

        output.writeLine("Average: %.2f MB/s", avg);

        return new MemoryBenchmarkResult() {
            {
                Average = avg;
                Output = output.Output;
            }
        };
    }

    protected SimpleEntry<Double, Integer> measureArrayRandomRead(int size) {
        int I = 0;
        int[] array = randomIntArray(size, Integer.MAX_VALUE);

        int end = array.length - 1;
        int[] indexes = randomIntArray(end, end);
        int k1 = (size / 1024) == 0 ? 1 : (size / 1024);
        int iterInternal = Iterrations / k1;
        for (int idx = 0; idx < end; idx++) {
            I = array[idx];
        }
        long start = System.currentTimeMillis();
        for (int i = 0; i < iterInternal; i++) {
            for (int idx : indexes) {
                I = array[idx];
            }
        }
        long elapsed = System.currentTimeMillis() - start;

        return new SimpleEntry<Double, Integer>(
                (double) iterInternal * array.length * 4 / (elapsed / 1000.0) / 1024 / 1024, I);
    }

    protected SimpleEntry<Double, Long> measureArrayRandomLongRead(int size) {
        long L = 0;
        long[] array = randomLongArray(size);

        int end = array.length - 1;
        int[] indexes = randomIntArray(end, end);
        int k1 = (size / 1024) == 0 ? 1 : (size / 1024);
        int iterInternal = Iterrations / k1;
        for (int idx = 0; idx < end; idx++) {
            L = array[idx];
        }
        long start = System.currentTimeMillis();
        for (int i = 0; i < iterInternal; i++) {
            for (int idx : indexes) {
                L = array[idx];
            }
        }
        long elapsed = System.currentTimeMillis() - start;

        return new SimpleEntry<Double, Long>(
                (double) iterInternal * array.length * 8 / (elapsed / 1000.0) / 1024 / 1024, L);
    }

}

File: src/dotnet/EntityFX.NetBenchamarks.Core/Generic/RandomMemoryBenchmarkBase.cs
using System;
using System.Linq;
using System.Diagnostics;
using System.Collections.Generic;

namespace EntityFX.NetBenchmark.Core.Generic
{

    public abstract class RandomMemoryBenchmarkBase<TResult> : BenchmarkBase<TResult>
    {

#if PocketPC
        private Dictionary<string, int> intMemTests = new Dictionary<string, int>
        {
            {"1k", 1024 / sizeof(int)}, {"32k", 32 * 1024 / sizeof(int)}, {"256k", 256 * 1024 / sizeof(int)}
        };

        private Dictionary<string, int> longMemTests = new Dictionary<string, int>
        {
            {"1k", 1024 / sizeof(long)}, {"32k", 32 * 1024 / sizeof(long)}, {"256k", 256 * 1024 / sizeof(long)}
        };
#else
        Dictionary<string, int> intMemTests = new Dictionary<string, int>
        {
            {"4k", 1024}, {"512k", 131072}, {"8M", 2097152}
        };

        Dictionary<string, int> longMemTests = new Dictionary<string, int>
        {
            {"4k", 512}, {"512k", 65536}, {"8M", 1048576}
        };
#endif

        private Random random;

        private IWriter localWriter;


        protected bool UseConsole { get { return writer.UseConsole; } set { writer.UseConsole = value; } }

        public RandomMemoryBenchmarkBase(bool printToConsole, IWriter writer)
            :base(writer)
        {
            localWriter = new Writer(null);
            Iterrations = 500000;
            Ratio = 2;
            random = new Random((int)DateTime.Now.Ticks);
            UseConsole = printToConsole;
        }

        public override void Warmup(double aspect)
        {
            UseConsole = false;
            base.Warmup(aspect);
            UseConsole = true;
        }

        public MemoryBenchmarkResult BenchRandomMemory()
        {
            double[] results = new double[intMemTests.Count + longMemTests.Count];

            int idx = 0;
            foreach (var item in intMemTests)
            {
                var res = MeasureArrayRandomRead(item.Value);
                results[idx] = res.Item1;
                localWriter.WriteLine(string.Format("Random int {0}: {1} MB/s", item.Key, res.Item1.ToString("F2")));

                idx++;
            }

            foreach (var item in longMemTests)
            {
                var res = MeasureArrayRandomLongRead(item.Value);
                results[idx] = res.Item1;
                localWriter.WriteLine(string.Format("Random long {0}: {1} MB/s", item.Key, res.Item1.ToString("F2")));

                idx++;
            }

            var avg = results.Average();
            localWriter.WriteLine(string.Format("Average: {0} MB/s", avg.ToString("F2")));
            writer.Write(localWriter.Output);
            return new MemoryBenchmarkResult()
            {
                Average = avg,
                Output = localWriter.Output
            };
        }

        protected MemoryMeasureResult<double, int> MeasureArrayRandomRead(int size)
        {
            int I = 0;
            var sw = new Stopwatch();
            var array = Enumerable.Range(0, size).Select(r => random.Next(int.MinValue, int.MaxValue)).ToArray();
            var end = array.Length - 1;
            var indexes = Enumerable.Range(0, end).Select(r => random.Next(0, end)).ToArray();
            var k1 = (size / 1024) == 0 ? 1 : (size / 1024) ;
            var iterInternal = Iterrations / k1;         
            for (int idx = 0; idx < end; idx++)
            {
                I = array[idx];
            }
            sw.Start();
            for (int i = 0; i < iterInternal; i++)
            {
                foreach (var idx in indexes)
                {
                    I = array[idx];
                }
            }
            sw.Stop();

            return new MemoryMeasureResult<double, int>((long)iterInternal * array.Length * 4 / sw.Elapsed.TotalSeconds / 1024 / 1024, I);
        }

        protected MemoryMeasureResult<double, long> MeasureArrayRandomLongRead(int size)
        {
            long L = 0;
            var sw = new Stopwatch();
            var array = Enumerable.Range(0, size).Select(r => (long)random.Next(int.MinValue, int.MaxValue)).ToArray();
            var end = array.Length - 1;
            var indexes = Enumerable.Range(0, end).Select(r => random.Next(0, end)).ToArray();
            var k1 = (size / 1024) == 0 ? 1 : (size / 1024) ;
            var iterInternal = Iterrations / k1;         
            for (int idx = 0; idx < end; idx++)
            {
                L = array[idx];
            }
            sw.Start();
            for (int i = 0; i < iterInternal; i++)
            {
                foreach (var idx in indexes)
                {
                    L = array[idx];
                }
            }
            sw.Stop();

            return new MemoryMeasureResult<double, long>((long)iterInternal * array.Length * 8 / sw.Elapsed.TotalSeconds / 1024 / 1024, L);
        }
    }
}
